File: Neetcode/MergeSortedLL.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def mergeTwoLists(self, list1: Optional[ListNode], list2: Optional[ListNode]) -> Optional[ListNode]:
        head1 = list1
        head2 = list2
        while head1 is not None and head2 is not None:
            # print(head1.val, list1.val, list2.val)
            if head1.val >= head2.val:
                head1, head2 = head2, head1
            if head1.next and head1.next.val < head2.val:
                head1 = head1.next
            elif head1.next is not None:
                temp = head1.next
                head1.next = head2
                head1 = temp
            elif head1.next is None and head2 is not None:
                head1.next = head2
                break
        
        if list1 is None:
            return list2
        elif list2 is None:
            return list1
        elif list1.val >= list2.val:
            return list2
        else:
            return list1

File: Neetcode/MergeSortedLL.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* mergeTwoLists(ListNode* list1, ListNode* list2) {
        if (list1 == nullptr){
            return list2;
        }
        if (list2 == nullptr){
            return list1;
        }
        ListNode* temp = nullptr;
        if (list1->val > list2->val){
            swap(list1, list2);
        }
        ListNode* rethead = list1;
        while (list1->next != nullptr){
            if (list1->next->val > list2->val){
                temp = list1->next;
                list1->next = list2;
                list1 = temp;
                swap(list1, list2);
            }
            else{
                list1 = list1->next;
            }
        }
        if (list2 != nullptr){
            list1->next = list2;
        }
        return rethead;
    }
};
